### PR TITLE
Use status bar for retriever update

### LINKED ISSUE #397: Use status bar for script download progress
Instead of printing out the name of each script being downloaded, use a status bar to indicate the progress of downloading all scripts.

### PR DESCRIPTION
**Pull request by ghoshbishakh**
Fixes #397
Use a progress bar while scripts are being downloaded for retriever update

### CODE REVIEW

File: lib/repository.py [hunk 1]
```diff
@@ -81,7 +82,39 @@ def flush(self):
...
+        progress = 0
+        status = "error: progress var must be float\r\n"
+    if progress < 0:
+        progress = 0
+        status = "Halt...\r\n"
+    if progress >= 1:
+        progress = 1
+        status = "Done...\r\n"
+    block = int(round(barWidth*progress))
+    text = "\r[{0}] {1:.2f}%".format("#"*block + "-"*(barWidth-block),
```
Comment (henrykironde) [inline lib/repository.py]: @ghoshbishakh great work. just a few things to change.
I think we can add Downlaod progress ` text = "\rDownload Progress: [{0}] `

File: lib/repository.py [hunk 2]
```diff
@@ -81,7 +82,39 @@ def flush(self):
...
         splash.Hide()
         sys.stdout = sys.__stdout__
-    print "The retriever is up-to-date"
+    print "\nThe retriever is up-to-date"
+
+
+def update_progressbar(progress):
+    """Shows progressbar given progress in float
+    Example: 0.5 for 50%
+    """
```
Comment (henrykironde) [inline lib/repository.py]: In the doc string we can remove this `Example:` 0.5 for `50%`
or we wan instead explain how this comes up
 
 """Shows progressbar given progress in float

accepts a a number  between 0 and 1 and it's converted to a float.
a number under 0 represents a 'halt'.
a number at 1 or greater represents 100%
"""

**Comment by ghoshbishakh** [comment]
@henrykironde thank you for taking a look. :smile:  I am updating my PR.

**Comment by ghoshbishakh** [comment]
@henrykironde  I hope that is better

File: lib/repository.py [hunk 3]
```diff
@@ -81,7 +82,39 @@ def flush(self):
...
+    try:
+        rows, columns = os.popen('stty size', 'r').read().split()
+        barWidth = int(columns) - 20
+        if(not (barWidth > 1)):
+            barWidth = 20
+    except:
+        pass
+    status = ""
+    if isinstance(progress, int):
+        progress = float(progress)
```
Comment (henrykironde) [inline lib/repository.py]: how about we change the logic and assume that given a number just convert to float, and this case we may not need to check if its int.
so before line 102, we can just say ` progress = float(progress)`

File: lib/repository.py [hunk 4]
```diff
@@ -175,7 +212,7 @@ def run(self):
                 path_script_name = os.path.normpath(os.path.join(HOME_DIR, "scripts", script_name))
 
                 if not file_exists(path_script_name):
-                    print "Downloading script: " + script_name
+                    # print "Downloading script: " + script_name
```
Comment (henrykironde) [inline lib/repository.py]: lets remove this `# print "Downloading script: " + script_name`

**Comment by ghoshbishakh** [comment]
good call :smile:  will update.

**Comment by henrykironde** [comment]
Modify this to change the **length** of the progress bar
barWidth = 20

if we go with bar_length or barLenth , people reading the code later may understand easily since the comment is talking about length

**Comment by ghoshbishakh** [comment]
ok sure but actually I am dynamically setting the length of the progressbar later according to the console width. So I think its better to comment that as default bar width?

**Comment by henrykironde** [comment]
ok that works

File: lib/repository.py [hunk 5]
```diff
@@ -81,7 +82,41 @@ def flush(self):
     if graphical:
         splash.Hide()
         sys.stdout = sys.__stdout__
-    print "The retriever is up-to-date"
+    print "\nThe retriever is up-to-date"
+
+
+def update_progressbar(progress):
+    """Shows progressbar given progress in float
+    accepts a a number between 0 and 1 and it's converted to a float.
```
Comment (henrykironde) [inline lib/repository.py]: space after the first line

Comment (ghoshbishakh) [inline lib/repository.py]: sorry I did not get that

Comment (ghoshbishakh) [inline lib/repository.py]: ah in the docstring ok!

**Comment by ghoshbishakh** [comment]
Did I miss anything?

File: lib/repository.py [hunk 6]
```diff
@@ -81,7 +82,39 @@ def flush(self):
...
+    bar_length = 20
+    # Try to set the bar_length according to the console size
+    try:
+        rows, columns = os.popen('stty size', 'r').read().split()
+        bar_length = int(columns) - 35
+        if(not (bar_length > 1)):
+            bar_length = 20
+    except:
+        pass
+    status = ""
```
Comment (henrykironde) [inline lib/repository.py]: if we are creating `status`, then I guess we may include it in line   ` text = "\rDownload Progress:....` 
or we can remove the creation of `status`

File: lib/repository.py [hunk 7]
```diff
@@ -81,7 +82,39 @@ def flush(self):
...
+
+    accepts a a number between 0 and 1 and it's converted to a float.
+    a number under 0 represents a 'halt'.
+    a number at 1 or greater represents 100%
+    """
+    # Modify this to change the default length of the progress bar
+    bar_length = 20
+    # Try to set the bar_length according to the console size
+    try:
+        rows, columns = os.popen('stty size', 'r').read().split()
```
Comment (henrykironde) [inline lib/repository.py]: do we need `row`?

Comment (ghoshbishakh) [inline lib/repository.py]: @henrykironde actually that returns a tuple thats why.
I can take index 2 though like [1].
Should I do that?

**Comment by ghoshbishakh** [comment]
@henrykironde cat you take a look at it now?

**Comment by henrykironde** [comment]
ok

**Comment by ethanwhite** [comment]
Thanks for this @ghoshbishakh. Overall it looks great.

I have a couple of thoughts/recommendations/questions:

1. I think we can remove the `reporthook` function entirely and then get rid of the associated optional argument in the call to `urllib.urlretriever`.

2. The docstring for `update_progressbar` could cleaned up a little (for part an explanation of 
part of the cleanup see number 4 below). E.g.,

    ```
    """Show progressbar

    Takes a number between 0 and 1 to indicate progress from 0 to 100%.

    """
    ```

3. Is the initial declaration of `bar_length` there for cases when determining the
size of the terminal doesn't work properly? If so then it would be clearer to
include this in the `except`. If not, then could you explain a little more about
what it's accomplishing.

4. Since `update_progressbar` shouldn't be getting values < 0 or > 1 I think we can
probably get rid of lines 105-112, and 116-117 unless I'm
missing something. If you were thinking about checking for bad values we could use
an `assert` statement and throw an exception if the value isn't between 0 and 1.

5. I don't understand why it's necessary to call `update_progressbar` at both the
beginning and end of the main `for` loop. Could you explain a little be more about
why we need both of them.

**Comment by ghoshbishakh** [comment]
@ethanwhite thanks for the suggestions :smile: 

1.  I am removing the reporthook

2.  Updated docstring

3.  Yes it was there for cases where determination of console size fails. I have put it in the except.

4.  I was thinking that later may be we could use statuses like halt for any exception or like that. But yes its better to keep it simple

5. Actually the 0% is not displayed if we don't call it first once. So I will call it outside the for loop first then so that it isn't repeated uselessly?

**Comment by ghoshbishakh** [comment]
@ethanwhite I have pushed with the required changes

**Comment by henrykironde** [comment]
great work @ghoshbishakh

**Comment by ghoshbishakh** [comment]
@henrykironde  thanks! :smile: 
Can you guide me a bit here:
https://github.com/numfocus/gsoc/issues/105

**Pull request merged**